Permit `repack` command in Scalar clones

mjcheetham opened the pull request:
Currently when the `core.gvfs` setting is set, several commands are outright blocked from running. Some of these commands, namely `repack` are actually OK to run in a Scalar clone, even if it uses the GVFS protocol (for Azure DevOps).

Introduce a different blocking mechanism to only block commands when the virtual filesystem is being used, rather than as a broad block on any `core.gvfs` setting.

mjcheetham committed:
git.c: add VFS enabled cmd blocking

Add the ability to block built-in commands based on if the `core.gvfs`
setting has the `GVFS_USE_VIRTUAL_FILESYSTEM` bit set. This allows us
to selectively block commands that use the GVFS protocol, but don't use
VFS for Git (for example repos cloned via `scalar clone` against Azure
DevOps).

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
git.c: permit repack cmd in Scalar repos

Loosen the blocking of the `repack` command from all "GVFS repos" (those
that have `core.gvfs` set) to only those that actually use the virtual
file system (VFS for Git only). This allows for `repack` to be used in
Scalar clones.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
worktree: remove special case GVFS cmd blocking

Replace the special casing of the `worktree` command being blocked on
VFS-enabled repos with the new `BLOCK_ON_VFS_ENABLED` flag.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

dscho reviewed (dismissed):
Nice!

dscho commented:
For the record: I believe that the original block to prevent `git repack` upon `core.gvfs` is rooted in the fact that VFS for Git deceives Git to consider all Git objects to be present locally, which would have caused, say, `git repack -adf` to download turn any VFS for Git-enabled clone into a full clone (which is obviously undesirable).

Scalar, on the other hand, initializes the clone as a partial clone, where Git is fully aware that it lacks some Git objects locally and does not try to fetch them during a `git repack`. Hence, that block makes little sense in Scalar-enabled clones and this PR is the right thing to do. @derrickstolee as the smart brain behind Scalar, you are much more familiar with this space than I am, do you agree, or did I make a mental mistake in this reasoning?

derrickstolee commented on git.c:
```diff
@@ -724,7 +724,7 @@ static struct cmd_struct commands[] = {
 	{ "version", cmd_version },
 	{ "whatchanged", cmd_whatchanged, RUN_SETUP },
-	{ "worktree", cmd_worktree, RUN_SETUP },
+	{ "worktree", cmd_worktree, RUN_SETUP | BLOCK_ON_VFS_ENABLED },
```
This is a much simpler way to do this!

derrickstolee commented on t/t0402-block-command-on-gvfs.sh:
```diff
@@ -36,4 +35,14 @@ test_expect_success 'test gc --auto succeeds when disabled via config' '
+	test_must_fail git repack
+'
+
+test_expect_success 'test repack succeeds with VFS bit disabled' '
```
Could you add a test with a Scalar clone that has a `gvfs.sharedCache` enabled? I'm curious as to whether `git repack` will operate on that object directory (my guess is not because its an alternate which Git treats as read-only).

It would be good to be up front about "`git repack` is no longer blocked, but it also won't do what you expect".

dscho replied:
> Could you add a test with a Scalar clone that has a `gvfs.sharedCache` enabled? I'm curious as to whether `git repack` will operate on that object directory (my guess is not because its an alternate which Git treats as read-only).

We do have this in `git maintenance run`: https://github.com/microsoft/git/blob/5ad6dd677514983f85322d9f7e48e1c58c1498e3/builtin/gc.c#L1822-L1831

I _think_ this is part of what is verified in https://github.com/microsoft/git/blob/5ad6dd677514983f85322d9f7e48e1c58c1498e3/t/t7900-maintenance.sh#L1126-L1155

mjcheetham replied:
The `cache-local-objects` maintenance task migrates local loose objects and packfiles to the shared cache. Also the `incremental-repack` task is redirected to the cache directory, so together the _maintenance tasks_ will repack the packs in the cache dir.

The `repack` built-in however will only act on the local objects. Would it be worth either:
a) reach `repack` to redirect to the shared object dir _instead_ (if set, like the `incremental-repack` task), or
b) print the warning, perhaps recommending a run of `git maintenance run --task=incremental-repack` instead?

mjcheetham replied:
> I'm curious as to whether `git repack` will operate on that object directory (my guess is not because its an alternate which Git treats as read-only).

There is an example of a similar scenario in `repack` already, whereby if an alternate has been configured the bitmaps are not written since "some objects are not being packed".

https://github.com/microsoft/git/blob/5ad6dd677514983f85322d9f7e48e1c58c1498e3/builtin/repack.c#L1292-L1302

dscho replied:
Hmm. If I put myself into a user's shoes, I would much rather prefer `git repack` to take care of the shared object directory (and potentially migrate objects there as part of the `repack`) if I work on a `gvfs.sharedCache`-enabled repository.

mjcheetham replied:
Something like this? (for option b)

```diff
diff --git a/builtin/repack.c b/builtin/repack.c
index 01db1b8f529..5a44171f8e8 100644
--- a/builtin/repack.c
+++ b/builtin/repack.c
@@ -24,6 +24,7 @@
 #include "pack-bitmap.h"
 #include "refs.h"
 #include "list-objects-filter-options.h"
+#include "gvfs.h"
 
 #define ALL_INTO_ONE 1
 #define LOOSEN_UNREACHABLE 2
@@ -1177,6 +1178,7 @@ int cmd_repack(int argc,
 	struct tempfile *refs_snapshot = NULL;
 	int i, ext, ret;
 	int show_progress;
+	const char *tmp_obj_dir = NULL;
 
 	/* variables to be filled by option parsing */
 	int delete_redundant = 0;
@@ -1301,6 +1303,10 @@ int cmd_repack(int argc,
 		write_bitmaps = 0;
 	}
 
+	if (gvfs_config_is_set(NULL) &&
+	    !git_config_get_value("gvfs.sharedcache", &tmp_obj_dir))
+		warning(_("shared object cache will not be repacked; use `git maintenance run --task=incremental-repack` instead"));
+
 	if (write_midx && write_bitmaps) {
 		struct strbuf path = STRBUF_INIT;
 

```

dscho replied:
That would be the option with the warning. However, as a user I'd much prefer the command to not warn me about something I probably don't want, but to do what I probably want instead, i.e. the other option you suggested ;-)

derrickstolee replied:
I think users who are bothering to run `git repack` themselves really want their object data to be reworked. I'm thinking specifically in the cases like `git repack -adf` that I use to rewrite all objects into a single packfile. We have historically avoided doing that on the shared object cache, but it should work somewhat well due to the integration with partial clone machinery. It's just going to be slow and users should expect that. We're not choosing to run it in the background because of that cost.

derrickstolee commented on git.c:
```diff
@@ -542,6 +543,9 @@ static int run_builtin(struct cmd_struct *p, int argc, const char **argv, struct
 	if (!help && p->option & BLOCK_ON_GVFS_REPO && gvfs_config_is_set(GVFS_BLOCK_COMMANDS))
 		die("'git %s' is not supported on a GVFS repo", p->cmd);
 
+	if (!help && p->option & BLOCK_ON_VFS_ENABLED && gvfs_config_is_set(GVFS_USE_VIRTUAL_FILESYSTEM))
```
This is really cool. Simplifies a lot!

derrickstolee reviewed (commented):
This seems to make sense, but also it may be good to add a warning in `git repack` when `gvfs.sharedCache` is enabled that it won't shrink the shared cache. What instructions should we add to do that full repack? Typically, I tell people to try `git maintenance run --task=incremental-repack` as the maintenance builtin knows how to adjust for the alternate _and_ that task doesn't care about refs. `git repack` with `GIT_OBJECT_DIRECTORY=<cache>` may struggle if there are local objects in the `.git/objects` dir that have not migrated to the cache.

mjcheetham commented:
> Scalar, on the other hand, initializes the clone as a partial clone, where Git is fully aware that it lacks some Git objects locally and does not try to fetch them during a `git repack`.

I guess we could also permit the `fsck` and `prune` commands in Scalar clones too, given that these are partial clone aware? cc @derrickstolee

mjcheetham committed:
git.c: permit fsck cmd in Scalar repos

Loosen the blocking of the `fsck` command from all "GVFS repos" (those
that have `core.gvfs` set) to only those that actually use the virtual
file system (VFS for Git only). This allows for `fsck` to be used in
Scalar clones.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham committed:
git.c: permit prune cmd in Scalar repos

Loosen the blocking of the `prune` command from all "GVFS repos" (those
that have `core.gvfs` set) to only those that actually use the virtual
file system (VFS for Git only). This allows for `prune` to be used in
Scalar clones.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

dscho reviewed (dismissed):
I guess this should not be retargeted to `vfs-2.49.0`, right? The changes look correct to me.

derrickstolee reviewed (commented):
I'm also prepared to approve this when it is retargeted. We can consider extensions to let folks repack their shared object cache as a follow-up.

mjcheetham committed:
builtin/repack.c: emit warning when shared cache is present

Emit a warning message when the `gvfs.sharedCache` option is set that
the `repack` command will not perform repacking on the shared cache.

In the future we can teach `repack` to operate on the shared cache, at
which point we can drop this commit.

Signed-off-by: Matthew John Cheetham <mjcheetham@outlook.com>

mjcheetham commented:
**Changes**

* Retargeted to `vfs-2.49.0`
* Unblock `prune` and `fsck` on Scalar repos
* Emit a warning when `gvfs.sharedCache` is set during `repack`

dscho commented on builtin/repack.c:
```diff
@@ -1301,6 +1303,10 @@ int cmd_repack(int argc,
 
+	if (gvfs_config_is_set(NULL) &&
+	    !git_config_get_value("gvfs.sharedcache", &tmp_obj_dir))
+		warning(_("shared object cache is configured but will not be repacked"));
```
Let's turn this warning into in informational message that the shared cache is repacked instead (and then do that) in a follow-up PR (and add a test to verify that it worked as intended). Unfortunately, it won't be as easy as:

```diff
-		warning(_("shared object cache is configured but will not be repacked"));
+		setenv(DB_ENVIRONMENT, tmp_obj_dir, 1);
```

because subsequent lines call `repo_get_object_repository()` ([this](https://github.com/microsoft/git/blob/3ca0cb9c52dc57366ddada2655a0fe875bb3a89a/builtin/repack.c#L1313) line, [this](https://github.com/microsoft/git/blob/3ca0cb9c52dc57366ddada2655a0fe875bb3a89a/builtin/repack.c#L1322) one and [this](https://github.com/microsoft/git/blob/3ca0cb9c52dc57366ddada2655a0fe875bb3a89a/builtin/repack.c#L1595) one), and that function [merely returns the already-initialized `repo->objects->odb->path`](https://github.com/microsoft/git/blob/48297853d4daddf6091cd2389a2294cda23ebfe1/repository.c#L108-L113).

Other than that, the [`pack-objects` command is spawned](https://github.com/microsoft/git/blob/3ca0cb9c52dc57366ddada2655a0fe875bb3a89a/builtin/repack.c#L1416) which also would need to know about this, and the cruft pack/delete redundant part probably also would need to be updated.

Most importantly, though, we really need to ensure that no objects are deleted that are reachable in _any_ Scalar repository that shares that `sharedcache` repository, even if the objects should not be reachable in that `sharedcache` repository itself...

dscho reviewed (dismissed):
Thank you!

After looking at the shared cache repacking idea briefly, I agree that it will be much safer to leave this for a separate PR.

derrickstolee reviewed (dismissed):
LGTM. Thanks for starting this effort!
